worldedit-core/src/main/java/com/fastasyncworldedit/core/history/changeset/AbstractChangeSet.java

```diff
@@ -166,38 +167,75 @@ public synchronized IChunkSet processSet(IChunk chunk, IChunkGet get, IChunkSet
             if (!set.hasSection(layer)) {
                 continue;
             }
-            // add each block and tile
-            char[] blocksGet;
-            char[] tmp = get.load(layer);
-            if (tmp == null) {
-                blocksGet = FaweCache.INSTANCE.EMPTY_CHAR_4096;
-            } else {
-                System.arraycopy(tmp, 0, (blocksGet = new char[4096]), 0, 4096);
-            }
-            char[] blocksSet;
-            // loadIfPresent shouldn't be null if set.hasSection(layer) is true
-            System.arraycopy(Objects.requireNonNull(set.loadIfPresent(layer)), 0, (blocksSet = new char[4096]), 0, 4096);
-
-            // Account for negative layers
-            int by = layer << 4;
-            for (int y = 0, index = 0; y < 16; y++) {
-                int yy = y + by;
-                for (int z = 0; z < 16; z++) {
-                    int zz = z + bz;
-                    for (int x = 0; x < 16; x++, index++) {
-                        int xx = bx + x;
-                        int from = blocksGet[index];
-                        if (from == BlockTypesCache.ReservedIDs.__RESERVED__) {
-                            from = BlockTypesCache.ReservedIDs.AIR;
+
+            if(!(get instanceof IntGetBlocks)) {
+                // add each block and tile
+                char[] blocksGet;
+                char[] tmp = get.loadChars(layer);
+                if (tmp == null) {
+                    blocksGet = FaweCache.INSTANCE.EMPTY_CHAR_4096;
+                } else {
+                    System.arraycopy(tmp, 0, (blocksGet = new char[4096]), 0, 4096);
+                }
+                char[] blocksSet;
+                // loadIfPresent shouldn't be null if set.hasSection(layer) is true
+                System.arraycopy(Objects.requireNonNull(set.loadCharsIfPresent(layer)), 0, (blocksSet = new char[4096]), 0, 4096);
+
+                // Account for negative layers
+                int by = layer << 4;
```

Comment:
Should be refactored to a private method, given it's copy/pasted from the `if` block.

Reply:
Refactored in next commit.